### ISSUE TITLE
FFmpeg windows builds will be removed

### ISSUE THREAD

**Issue by rom1v** [Genymobile/scrcpy#1753]
I developed scrcpy for Linux, where the dependencies are just retrieved via `apt`.

But scrcpy had to work on Windows too. I'm bad at build systems, and have a very limited knowledge of Windows, so I initially just recommended to build from something which looks like a Linux environment (MSYS2). For the first public release (v1.0), I even just took the DLL installed from the MSYS environment.

But it was too cumbersome for me to create a release (I had to build from a Windows machine), so I wrote some script to cross-compile the binaries from Linux, and just downloaded prebuilt dependencies of FFmpeg and SDL (106b87a4d2459ae881ab89b987cc9347a843a4ea). This was a quick&dirty solution (which btw makes adding or upgrading dependencies very annoying). I'm not very proud of it, but at least it was good enough for solving my problem.

The FFmpeg binaries are downloaded from https://ffmpeg.zeranoe.com/builds/. This site will close soon:

> ffmpeg.zeranoe.com will close on Sep 18, 2020, and all builds will be removed. If you're using Zeranoe FFmpeg Builds in your product, please ensure that you have a hosting strategy in place by the 18th. Thank you to everyone who supported the project along the way!

so the release script of scrcpy will just stop to work correctly.

In the short term, I could just keep the latest versions and ship them. But it becomes urgent to build the dependencies along with scrcpy. And an option to build static binaries would be awesome (#1624).

https://linuxfr.org/users/roger21/journaux/ffmpeg-pour-windows-ca-va-couper

**Comment by npes87184** [comment]
#1484 also suffered from dll hell. IMO, it can be solved in the same time?

**Comment by Biswa96** [comment]
> I even just took the DLL installed from the MSYS environment.

MSYS2/MinGW FFmpeg builds are dynamically linked. So, you have to download all the dependencies which may not be the right solution here.

> The FFmpeg binaries are downloaded from https://ffmpeg.zeranoe.com/builds/. This site will close soon:

How those guys build it. Any script?

> so the release script of scrcpy will just stop to work correctly.

Neevvveeerrr. With CI, possibility is endless.

One lazy solution would be to create a package for MSYS2 which already provides FFmpeg and SDL2. But the package will not be a standalone like the current releases.

**Comment by ouvtk** [comment]
Not sure if it will help much, but ffmpeg peeps already have a compilation guides: https://trac.ffmpeg.org/wiki/CompilationGuide
And https://github.com/rdp/ffmpeg-windows-build-helpers looks as the best option for cross-compiling.

But need to understand what the end goal would be, as compiling ffmpeg every as a dependency for scrcpy would be time consuming.

**Comment by rom1v** [comment]
> MSYS2/MinGW FFmpeg builds are dynamically linked. So, you have to download all the dependencies which may not be the right solution here.

Yes, that's what I did for v1.0. :smile:

> How those guys build it. Any script?

I don't know.

In theory, I'd like to do something similar as we do in VLC (cf [`contrib/src/ffmpeg/rules.mak`](https://code.videolan.org/videolan/vlc/-/blob/master/contrib/src/ffmpeg/rules.mak)).

> Neevvveeerrr. With CI, possibility is endless.

Yes, but I would like to be able to build the whole release from any computer (running linux) in a single script (even without docker) :)

**Comment by rom1v** [comment]
> Not sure if it will help much, but ffmpeg peeps already have a compilation guides: https://trac.ffmpeg.org/wiki/CompilationGuide
And https://github.com/rdp/ffmpeg-windows-build-helpers looks as the best option for cross-compiling.

Thank you for the links :+1:

> as compiling ffmpeg every as a dependency for scrcpy would be time consuming

Only if we want/need to, otherwise it could use the dependencies provided by the system.

But it would also allow to create a minimal static build (that would be awesome to distribute the latest scrcpy release on linux).

**Comment by Biswa96** [comment]
> In theory, I'd like to do something similar as we do in VLC

If that need a script to be added in scrcpy, would you add it?

> Yes, but I would like to be able to build the whole release from any computer

By whole release, do you mean to compile dependencies also? Or scrcpy only?

**Comment by rom1v** [comment]
> If that need a script to be added in scrcpy, would you add it?

Yes, a script which builds ffmpeg and a script which builds SDL. :+1:

I would like to replace `prebuild-deps` by a directory `deps` with 3 subdirs:
 - `deps/src/ffmpeg`
 - `deps/src/sdl`
 - `deps/src/adb`

Each subdir would contain a script to build and install the dependency into some destination directory (or just download and verify checksum for `adb`). The resulting built libraries would be used by the scrcpy build system.

But that should be optional (e.g. we should be able to build scrcpy against the system ffmpeg).

> By whole release, do you mean to compile dependencies also? Or scrcpy only?

Yes, dependencies included.

The release script should generate the files posted on every release:
 - scrcpy-server
 - windows 32 bits release
 - windows 64 bits release

If we compile the dependencies ourselves, either we just include the dll in the release, or better we build the scrcpy binaries with FFmpeg and libsdl statically linked.

And if that's the case, we could also generate:
 - linux 32 bits release
 - linux 64 bits release

This is just what I have currently in mind, maybe there are better ways.

**Comment by Biswa96** [comment]
Found something https://github.com/rdp/ffmpeg-windows-build-helpers.git. I will try it.

**Comment by npes87184** [comment]
> Found something https://github.com/rdp/ffmpeg-windows-build-helpers.git. I will try it.

Be aware that the license of this script is GPLv3.

**Comment by npes87184** [comment]
Hi, @rom1v 

Would you accept something like [mxe](https://github.com/mxe/mxe)? I built SDL2 and ffmpeg using mxe. After it, I built scrcpy against the output dlls in mxe and copied needed dll by using [copydlldeps](https://github.com/mxe/mxe/blob/master/tools/copydlldeps.md) (a tool from mxe).

**Comment by rom1v** [comment]
> Would you accept something like mxe?

For now, I don't know enough to have an opinion.

> MXE (M cross environment) is a GNU Makefile that compiles a cross compiler and cross compiles many free libraries such as SDL and Qt.

So compiling the cross compiler is a necessary step?

**Comment by rom1v** [comment]
Oh, ffmpeg.org provides an alternative link to builds of FFmpeg:
 - http://ffmpeg.org/download.html#build-windows
 - https://github.com/BtbN/FFmpeg-Builds/releases

This does not change the goal to get a proper solution, but it could help in the short term :)

**Comment by npes87184** [comment]
> > Would you accept something like mxe?
> 
> For now, I don't know enough to have an opinion.

For demo and testing, I can provide a PoC in a branch of my fork repo.

> 
> > MXE (M cross environment) is a GNU Makefile that compiles a cross compiler and cross compiles many free libraries such as SDL and Qt.
> 
> So compiling the cross compiler is a necessary step?

From my point of view, the answer looks yes.

**Comment by npes87184** [comment]
Hi, @rom1v 

You can try this [commit](https://github.com/npes87184/scrcpy/commit/c83f60e46590d328aa46f4df5ce75a7c67eb246f) if you have an interest in this.

I just simply modified the win64 version.

Thanks.

**Comment by rom1v** [comment]
@npes87184 Oh, thank you! Very interesting.

It looks great in the Makefile.

However, since it rebuilds everything (including gcc and other stuff), it takes too much time. :disappointed:

**Comment by npes87184** [comment]
> @npes87184 Oh, thank you! Very interesting.
> 
> It looks great in the Makefile.

I guess it still has some problems need to be solved. When it executes the `copydlldeps.sh`, it shows that some dlls does not find. Even though, I can run this in my windows 10, I am not sure if it affects or not.

> 
> However, since it rebuilds everything (including gcc and other stuff), it takes too much time. 😞

It also has pre-built libs [here](https://pkg.mxe.cc/pkg/). However, I have not tested it.

BTW, if we want to do static build, we need to write down all libs that linked by ffmpeg and SDL2. A little bit annoying.

**Comment by rom1v** [comment]
> BTW, if we want to do static link ffmpeg, it also needs to care about the license.

True. 

Static link is a nice-to-have solution so that Linux users could quickly test the latest version using a binary. If we are able to build the deps dynamically for Windows, it's already good.

**Comment by npes87184** [comment]
> Oh, ffmpeg.org provides an alternative link to builds of FFmpeg:
> 
> * http://ffmpeg.org/download.html#build-windows
> * https://github.com/BtbN/FFmpeg-Builds/releases
> 
> This does not change the goal to get a proper solution, but it could help in the short term :)

I create a PR that use ffmpeg built from BtbN in #1838 .

**Comment by deltachonglee81** [comment]
Save and modify the system

**Comment by deep9** [comment]
Until this is resolved, you can prepend the ffmpeg url with https://web.archive.org/web/20200916091820/ (like https://web.archive.org/web/20200916091820/https://ffmpeg.zeranoe.com/builds/win32/shared/ffmpeg-4.3.1-win32-shared.zip) in prebuild-deps/Makefile

**Comment by rom1v** [comment]
@deep9 Thank you for your suggestion. I tested, but their server are quite slow for downloading such archives (and I guess it's not very fair to download from them).

Instead, I added the releases on GitHub, and referenced them: 5dc3285dbf807553f86346ab16d0891cfeb32ab7

**Comment by deep9** [comment]
@rom1v Sure, this is more stable solution, mine was intended as a workaround :)

**Comment by LHLaurini** [comment]
May I suggest using [Conan](https://conan.io/)? It's able to both download prebuilt binaries (if available) or build automatically from the source. I haven't used it much, but it's worked well for me so far.

**Comment by rom1v** [comment]
> Problem is it's much better integrated with CMake. Meson, not so much.

Note: it seems that meson [can use the CMake `find_package`](https://mesonbuild.com/Release-notes-for-0-49-0.html#cmake-find_package-dependency-backend) for compatibility.

(However, Conan is not available is official Debian repositories, so I prefer not to depend on it.)

**Comment by Menstro** [comment]
Unsubscribe

**Comment by rom1v** [comment]
@Menstro

> Unsubscribe

To unsubscribe from a thread on github, there is a button `Unsubscribe` on the right column, in the "Notifications" section.

**Comment by LHLaurini** [comment]
> (However, Conan is not available is official Debian repositories, so I prefer not to depend on it.)

Conan can be installed using `pip`.

Even then, seems [the official Conan repo doesn't have a ffmpeg recipe yet](https://github.com/conan-io/conan-center-index/issues/4983). Maybe once it's available, we could give it another thought.

**Comment by MartinDelille** [comment]
> Even then, seems the official Conan repo doesn't have a ffmpeg recipe yet. Maybe once it's available, we could give it another thought.

It's available and functional: https://github.com/conan-io/conan-center-index/pull/6707

**Comment by AlynxZhou** [comment]
If we need to build dependencies on windows, I suggest to try meson's wrap, which can helps to gather and build dependencies. See <https://github.com/Genymobile/scrcpy/issues/1624#issuecomment-918846360>.

**Comment by AlynxZhou** [comment]
However wrapdb has SDL2 but not ffmpeg.

**Comment by AlynxZhou** [comment]
A meson-ported ffmpeg hosted by gstreamer is <https://gitlab.freedesktop.org/gstreamer/meson-ports/ffmpeg> but it's ffmpeg 4.1.

**Comment by rom1v** [comment]
I published a custom build specific to scrcpy (containing only the features used by scrcpy):
https://github.com/rom1v/scrcpy-deps

Here is a commit (not merged yet) to use it: https://github.com/Genymobile/scrcpy/commit/custom-ffmpeg

For comparison, here are the sizes of the dll for FFmpeg 5.1.2:
 - before: 89M
 - after: 4.7M

**Comment by rom1v** [comment]
And now, [FFmpeg 6.0](https://www.ffmpeg.org/download.html#release_6.0) :tada: 

https://github.com/rom1v/scrcpy-deps/releases/tag/6.0-scrcpy
